models/cross-spork_client.go

```diff
@@ -0,0 +1,124 @@
+type CrossSporkClient struct {
+	logger zerolog.Logger
+	// this map holds the last heights and clients for each spork
+	sporkHosts map[uint64]access.Client
+
+	access.Client
+}
+
+// NewCrossSporkClient creates a new instance of the client, it accepts the
+// host to the current spork AN API.
+func NewCrossSporkClient(currentSporkHost string, logger zerolog.Logger) (*CrossSporkClient, error) {
+	// add current spork AN host as the default client
+	client, err := grpc.NewClient(currentSporkHost)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CrossSporkClient{
+		logger,
+		make(map[uint64]access.Client),
+		client,
+	}, nil
+}
+
+func (c *CrossSporkClient) AddSpork(lastHeight uint64, host string) error {
+	if _, ok := c.sporkHosts[lastHeight]; ok {
+		return fmt.Errorf("provided last height already exists")
+	}
+
+	client, err := grpc.NewClient(host)
+	if err != nil {
+		return err
+	}
+
+	c.sporkHosts[lastHeight] = client
+
+	c.logger.Info().
+		Uint64("spork-boundary", lastHeight).
+		Str("host", host).
+		Msg("added spork specific client")
+
+	return nil
+}
+
+// getClientForHeight returns the client for the given height. It starts by using the current spork client,
+// then iteratively checks the upper height boundaries in descending order and returns the last client
+// that still contains the given height within its upper height limit. If no client is found, it returns
+// the current spork client.
+// Please note that even if a client for provided height is found we don't guarantee the data being available
+// because it still might not have access to the height provided, because there might be other sporks with
+// lower height boundaries that we didn't configure for.
+// This would result in the error when using the client to access such data.
+func (c *CrossSporkClient) getClientForHeight(height uint64) access.Client {
+	heights := maps.Keys(c.sporkHosts)
+	slices.Sort(heights)    // order heights in ascending order
+	slices.Reverse(heights) // make it descending
+
+	// start by using the current spork client, then iterate all the upper height boundaries
+	// and find the last client that still contains the height in its upper height limit
+	client := c.Client
+	for _, upperBound := range heights {
+		if upperBound >= height {
+			client = c.sporkHosts[upperBound]
+
+			c.logger.Debug().
+				Uint64("spork-boundary", upperBound).
+				Msg("using previous spork client")
+		}
+	}
+
+	return client
+}
+
+func (c *CrossSporkClient) GetBlockByHeight(
+	ctx context.Context,
+	height uint64,
+) (*flow.Block, error) {
+	return c.
+		getClientForHeight(height).
+		GetBlockByHeight(ctx, height)
+}
+
+func (c *CrossSporkClient) ExecuteScriptAtBlockHeight(
+	ctx context.Context,
+	height uint64,
+	script []byte,
+	arguments []cadence.Value,
+) (cadence.Value, error) {
+	return c.
+		getClientForHeight(height).
+		ExecuteScriptAtBlockHeight(ctx, height, script, arguments)
+}
+
+func (c *CrossSporkClient) SubscribeEventsByBlockHeight(
```

Comment:
what should the behavior be when the last block is reached? The AN streaming API doesn't handle the case where the spork ends, so it will just pause indefinitely once the last sealed block is reached.

Reply:
This I need to still add yeah, sorry, share a PR a bit early :D didn't think you will be this fast